src/ocrmypdf/_jobcontext.py

```diff
@@ -17,67 +17,120 @@
+        self.options = options
+        self.work_folder = work_folder
+        self.origin = origin
+        self.pdfinfo = pdfinfo
+        if options:
+            self.name = os.path.basename(options.input_file)
+        else:
+            self.name = 'origin.pdf'
+        if self.name == '-':
+            self.name = 'stdin'
+        self.log = get_logger(options, '%s: ' % self.name)
 
-    While this would naturally lend itself @property's it seems to make
-    a little more sense to use functions to make it explicitly that the
-    invocation requires marshalling data across a process boundary.
+    def get_path(self, name):
+        return os.path.join(self.work_folder, name)
 
-    """
+    def get_page_contexts(self):
+        npages = len(self.pdfinfo)
+        for n in range(npages):
+            yield PageContext(self, n)
 
-    def __init__(self):
-        self.pdfinfo = None
-        self.options = None
-        self.work_folder = None
-        self.rotations = {}
 
-    def generate_pdfinfo(self, infile):
-        self.pdfinfo = PdfInfo(infile)
+class PageContext:
+    """Holds our context for a page"""
 
-    def get_pdfinfo(self):
-        "What we know about the input PDF"
-        return self.pdfinfo
+    def __init__(self, pdf_context, pageno):
+        self.pdf_context = pdf_context
+        self.options = pdf_context.options
+        self.pageno = pageno
+        self.pageinfo = pdf_context.pdfinfo[pageno]
+        self.log = get_logger(pdf_context.options, '%s Page %d: ' % (pdf_context.name, pageno + 1))
 
-    def set_pdfinfo(self, pdfinfo):
-        self.pdfinfo = pdfinfo
+    def get_path(self, name):
+        return os.path.join(self.pdf_context.work_folder, "%06d_%s" % (self.pageno + 1, name))
 
-    def get_options(self):
-        return self.options
 
-    def set_options(self, options):
-        self.options = options
+def cleanup_working_files(work_folder, options):
+    if options.keep_temporary_files:
+        print(f"Temporary working files saved at:\n{work_folder}", file=sys.stderr)
+    else:
+        with suppress(FileNotFoundError):
+            shutil.rmtree(work_folder)
 
-    def get_work_folder(self):
-        return self.work_folder
 
-    def set_work_folder(self, work_folder):
-        self.work_folder = work_folder
+def get_logger(options=None, prefix=''):
+    level = ERROR  # TODO: add option
+    if options is not None:
+        if options.quiet or options.output_file == '-' or options.sidecar == '-':
+            return NullLogger()
+        if options.verbose == 0:
+            level = ERROR
+        elif options.verbose == 1:
+            level = WARN
+        elif options.verbose == 2:
+            level = INFO
+        elif options.verbose >= 3:
+            level = DEBUG
 
-    def get_rotation(self, pageno):
-        return self.rotations.get(pageno, 0)
+    return Logger(prefix, level)
 
-    def set_rotation(self, pageno, value):
-        self.rotations[pageno] = value
 
+class Logger:
```

Comment:
With ruffus out of the picture we can use the Python `logging` module instead of a custom logger.

If we need to use ProcessPoolExecutor instead there's a little setup involved but it should be quite doable.